File: app/src/main/java/com/abc/danielharrington/betterdays/SettingsFragment.java
package com.abc.danielharrington.betterdays;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Random;

import static android.content.Context.MODE_PRIVATE;


public class SettingsFragment extends PreferenceFragmentCompat {

    private ListPreference themePreference;
    private ListPreference notificationsPreference;
    private Preference savePreference;

    private String THEME_SELECTED;
    private int NOTIFICATIONS_PER_DAY;

    public static final String SHARED_PREFS = "sharedPrefs";
    public static final String THEME_PREF = "theme_preference";
    public static final String NOTS_PREF = "notifications_preference";

    private String theme;
    private int notifications;

    ArrayList<Calendar> calList;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.preferences);



        themePreference = findPreference("theme_preference");
        notificationsPreference = findPreference("notifications_preference");
        savePreference = findPreference("save_preference_button");

        calList = new ArrayList<>();

        //handling the save
        savePreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                saveData();

                return true;
            }
        });

        //handling the theme preference
        themePreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                THEME_SELECTED = newValue.toString(); //setting the selection to the constant for saving
                return true;
            }
        });

        //handling the notifications preference
        notificationsPreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                NOTIFICATIONS_PER_DAY = Integer.parseInt(newValue.toString()); //setting the selection to the constant for saving
                return true;
            }
        });

    }//onCreatePreferences

    //method to save preferences when the user clicks "SAVE"
    public void saveData(){

        if(NOTIFICATIONS_PER_DAY > 0) {
            setAlarms();
        }else{
            clearAlarms();
        }

        SharedPreferences sharedPreferences = getContext().getSharedPreferences(SHARED_PREFS, MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();

        editor.putString(THEME_PREF, THEME_SELECTED);
        editor.putInt(NOTS_PREF, NOTIFICATIONS_PER_DAY);

        editor.apply();

        Toast.makeText(getContext(), "Settings saved", Toast.LENGTH_SHORT).show();
    }//saveData method

    //method to load the saved data
    public void loadData(){
        SharedPreferences sharedPreferences = getContext().getSharedPreferences(SHARED_PREFS, MODE_PRIVATE);
        theme = sharedPreferences.getString(THEME_PREF, "Default");
        notifications = sharedPreferences.getInt(NOTS_PREF, 0);
    }//loadData method

    //method to set repeating notification alarms (random times)
    public void setAlarms(){
        Random rand = new Random();
        int hour;
        int minute;

        for (int i = 0; i < NOTIFICATIONS_PER_DAY; i++){
            hour = rand.nextInt(25);
            minute = rand.nextInt(61);
            Calendar cal = Calendar.getInstance();
            cal.set(Calendar.HOUR_OF_DAY, hour);
            cal.set(Calendar.MINUTE, minute);
            cal.set(Calendar.SECOND, 0);

            calList.add(cal);
        }//for

        for(Calendar cal: calList) {
            int i = 0;
            AlarmManager alarmManager = (AlarmManager) getContext().getSystemService(Context.ALARM_SERVICE);
            Intent intent = new Intent(getContext(), AlertReceiver.class);
            PendingIntent pendingIntent = PendingIntent.getBroadcast(getContext(), 1, intent, i);

            alarmManager.setInexactRepeating(AlarmManager.RTC_WAKEUP, cal.getTimeInMillis(), AlarmManager.INTERVAL_DAY, pendingIntent);
            i++;
        }//for
    }//setAlarms method

    //method to clear the alarms
    public void clearAlarms(){
        for(Calendar cal: calList) {
            int i = 0;
            AlarmManager alarmManager = (AlarmManager) getContext().getSystemService(Context.ALARM_SERVICE);
            Intent intent = new Intent(getContext(), AlertReceiver.class);
            PendingIntent pendingIntent = PendingIntent.getBroadcast(getContext(), 1, intent, i);

            alarmManager.cancel(pendingIntent);
            i++;
        }//for

    }//clearAlarms method

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        view.setBackgroundResource(R.color.signBlue);
    }
}//SettingsFragment class

File: app/src/main/java/com/abc/danielharrington/betterdays/SettingsFragment.kt
package com.abc.danielharrington.betterdays
/*
A good deal of the heavy lifting is being done in this fragment
alongside the main activity.
 */


import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import java.util.ArrayList
import java.util.Calendar
import android.content.Context.MODE_PRIVATE
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.PeriodicWorkRequest
import androidx.work.WorkManager
import java.util.concurrent.TimeUnit


class SettingsFragment : PreferenceFragmentCompat() {

    private var themePreference: ListPreference? = null
    private var notificationsPreference: ListPreference? = null
    private var savePreference: Preference? = null

    //private var NOTIFICATIONS_PER_DAY: Int = 0

    //private var calList: ArrayList<Calendar> = ArrayList()

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences, rootKey)

        themePreference = findPreference("theme_preference")
        notificationsPreference = findPreference("notifications_preference")
        savePreference = findPreference("save_preference_button")

        calList = ArrayList()

        //handling the theme preference
        themePreference!!.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { preference, newValue ->
            THEME_SELECTED = newValue.toString() //setting the selection to the constant for saving
            true
        }

        //handling the notifications preference
        notificationsPreference!!.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { preference, newValue ->
            NOTIFICATIONS_PER_DAY = Integer.parseInt(newValue.toString()) //setting the selection to the constant for saving
            true
        }

        //handling the save
        savePreference!!.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            saveData()
            true
        }

    }//onCreatePreferences

    //method to save preferences when the user clicks "SAVE"
    private fun saveData() {

        if (NOTIFICATIONS_PER_DAY > 0) {
            startWork() //starting periodic WorkManager
        } else {
            cancelWork() //cancelling the work if the user doesn't want notifications anymore
        }//if/else

        val sharedPreferences = activity!!.getSharedPreferences(SHARED_PREFS, MODE_PRIVATE)
        val editor = sharedPreferences.edit()

        editor.putString(THEME_PREF, THEME_SELECTED)
        editor.putInt(NOTS_PREF, NOTIFICATIONS_PER_DAY)

        editor.apply()

        Toast.makeText(context, "Settings saved", Toast.LENGTH_SHORT).show()
    }//saveData method

    private fun startWork(){
        //cancel any work that already exists
        cancelWork()

        var delayTime: Long

        for(i in 0 until NOTIFICATIONS_PER_DAY){
            delayTime = (0..23).random().toLong() //to mimic a "random" time, not taking minutes into account

            val task = PeriodicWorkRequest.Builder(
                   NotificationWorker::class.java,
                    1,
                    TimeUnit.DAYS,
                    30,
                    TimeUnit.MINUTES)
                    .addTag("periodic_quote_not")
                    .setInitialDelay(delayTime, TimeUnit.HOURS)
                    .build()
            println("Delay time: $delayTime")
            WorkManager.getInstance(context!!).enqueueUniquePeriodicWork("periodic_quote_work", ExistingPeriodicWorkPolicy.REPLACE, task)

        }//for

    }//startWork method

    private fun cancelWork(){
        WorkManager.getInstance(context!!).cancelAllWorkByTag("periodic_quote_not") //used one tag for all
    }//cancelWork method

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        view.setBackgroundResource(R.color.signBlue)
        activity!!.title = "Settings"
    }//onViewCreated method

    companion object {

        const val SHARED_PREFS = "sharedPrefs"
        const val THEME_PREF = "themePreference"
        const val NOTS_PREF = "notificationsPreference"
        const val QUOTES_PREF = "quotesListPreference"
        const val SPEAKERS_PREF = "speakersListPreference"
        const val SAVED_QUOTE = "savedQuote"
        const val SAVED_SPEAKER = "savedSpeaker"
        var THEME_SELECTED: String? = null
        var NOTIFICATIONS_PER_DAY: Int = 0
        var calList: ArrayList<Calendar> = ArrayList()
    }//companion object
}//SettingsFragment class